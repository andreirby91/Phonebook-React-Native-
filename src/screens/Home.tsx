import React, {useState, useEffect, useContext} from 'react';
import {StyleSheet, Text, View, FlatList, TouchableOpacity} from 'react-native';
import {StackNavigationProp} from '@react-navigation/stack';
import {RootStackParamList} from '../../App';
import ContactItem from '../components/ContactItem';
import Search from '../components/Search';
import EmptyContactList from '../components/EmptyContactList';
import AppStateContext from '../context/app.context';

type MainScreenNavigationProp = StackNavigationProp<RootStackParamList, 'Main'>;

type Props = {
  navigation: MainScreenNavigationProp;
};

export type Contact = {
  id: number;
  name: string;
  phone: string;
  sex: string;
  country: string;
  code: string;
};

const styles = StyleSheet.create({
  list: {
    paddingHorizontal: 10,
    marginVertical: 10,
  },
  separator: {
    borderBottomWidth: 1,
    borderColor: '#ddd',
    marginVertical: 10,
  },
  header: {
    alignItems: 'center',
    paddingTop: 20,
  },
  emptyList: {
    alignItems: 'center',
    justifyContent: 'center',
  },
});

const Home = ({navigation}: Props) => {
  const [searchValue, setSearchValue] = useState<string>('');
  const [myContacts, setMyContacts] = useState<Contact[]>([]);
  const [filteredContacts, setFilteredContacts] = useState<Contact[]>([]);
  const {contactsList, getContactsFromStorage, removeContact} = useContext(
    AppStateContext,
  );

  useEffect(() => {
    getContactsFromStorage();
  }, []);

  useEffect(() => {
    handleContactUpdates(contactsList);
  }, [contactsList]);

  const handleFilteringByType = (input: string) => {
    let isNumRegex = /^[0-9]+$/;
    const filterBy = isNumRegex.test(input) ? 'phone' : 'name';
    const filteredItems = myContacts.filter((contact: Contact) =>
      contact[filterBy].toLowerCase().includes(input.toLowerCase()),
    );

    return filteredItems;
  };

  const handleSearch = (text: any) => {
    setSearchValue(text);
    setFilteredContacts(handleFilteringByType(text));
  };

  const handleContactUpdates = (contacts: Contact[]) => {
    setMyContacts(contacts);
    setFilteredContacts(contacts);
  };

  const handleDelete = (contactId: number) => {
    removeContact(contactId);
  };

  const ListEmptyComponent = () => {
    return myContacts.length ? (
      <View style={styles.emptyList}>
        <Text>No contacts found!</Text>
      </View>
    ) : (
      <EmptyContactList onPress={() => navigation.navigate('AddNewContact')} />
    );
  };

  return (
    <>
      <View style={styles.header}>
        <TouchableOpacity onPress={() => navigation.navigate('AddNewContact')}>
          <Text>Add new contact</Text>
        </TouchableOpacity>
        <Search
          searchValue={searchValue}
          onChangeText={(text: string) => handleSearch(text)}
        />
      </View>
      <FlatList
        style={styles.list}
        data={filteredContacts}
        keyExtractor={(item) => item.id.toString()}
        renderItem={({item}) => {
          const {name, phone, sex, country, code, id} = item;
          return (
            <ContactItem
              name={name}
              phone={phone}
              sex={sex}
              country={country}
              code={code}
              onRightPress={() => handleDelete(id)}
            />
          );
        }}
        ItemSeparatorComponent={() => <View style={styles.separator}></View>}
        ListEmptyComponent={ListEmptyComponent}
      />
    </>
  );
};

export default Home;
