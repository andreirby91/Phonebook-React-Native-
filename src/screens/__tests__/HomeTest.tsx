import React from 'react';
import renderer from 'react-test-renderer';

import Home from '../Home';
import {NavigationContainer} from '@react-navigation/native';

it('renders correctly with defaults', () => {
  const button = renderer
    .create(
      <NavigationContainer>
        <Home />
      </NavigationContainer>,
    )
    .toJSON();
  expect(button).toMatchSnapshot();
});
