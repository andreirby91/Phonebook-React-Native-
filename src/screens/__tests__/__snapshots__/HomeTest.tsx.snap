// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with defaults 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingTop": 20,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      onChangeText={[Function]}
      placeholder="Search by names and numbers:"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#fafafb",
          "borderColor": "grey",
          "borderRadius": 20,
          "borderWidth": 0.5,
          "flex": 1,
          "fontSize": 18,
          "height": 45,
          "marginVertical": 10,
          "maxHeight": 45,
          "paddingHorizontal": 20,
          "paddingVertical": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#0984e3",
          "borderRadius": 30,
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "center",
          "marginLeft": 20,
          "minWidth": 40,
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 12,
            },
            Object {
              "color": "#fff",
              "fontSize": 18,
              "textAlign": "center",
            },
            Object {
              "fontFamily": "Fontisto",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 18,
          }
        }
      />
    </View>
  </View>,
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "marginVertical": 10,
        "paddingHorizontal": 10,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginTop": 50,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 12,
              },
              Object {
                "fontSize": 70,
                "marginBottom": 30,
              },
              Object {
                "fontFamily": "Fontisto",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text
          style={
            Object {
              "fontSize": 20,
              "marginBottom": 20,
              "textAlign": "center",
            }
          }
        >
          No contacts available!
        </Text>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#0984e3",
              "borderRadius": 30,
              "height": 50,
              "justifyContent": "center",
              "minWidth": 200,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#fff",
                "fontSize": 18,
              }
            }
          >
            Add new contact
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;
